src/algokit/core/dispenser.py

```diff
@@ -0,0 +1,361 @@
+    }
+
+    if method.upper() != "GET" and data is not None:
+        request_args["json"] = data
+
+    try:
+        response: httpx.Response = getattr(httpx, method.lower())(**request_args)
+        response.raise_for_status()
+        return response
+
+    except httpx.HTTPStatusError as err:
+        error_message = f"Error processing dispenser API request: {err.response.status_code}"
+
+        if err.response.status_code == httpx.codes.BAD_REQUEST:
+            error_message = err.response.json()["message"]
+
+        raise Exception(error_message) from err
+
+    except Exception as err:
+        error_message = "Error processing dispenser API request"
+        logger.debug(f"{error_message}: {err}", exc_info=True)
+        raise err
+
+
+def set_dispenser_credentials(token_data: dict[str, str]) -> None:
+    """
+    Set the keyring passwords.
+    """
+
+    decoded_id_token = jwt.decode(token_data["id_token"], algorithms=ALGORITHMS, options={"verify_signature": False})
+
+    keyring.set_password(DISPENSER_KEYRING_NAMESPACE, DISPENSER_KEYRING_ID_TOKEN_KEY, token_data["id_token"])
+    keyring.set_password(DISPENSER_KEYRING_NAMESPACE, DISPENSER_KEYRING_ACCESS_TOKEN_KEY, token_data["access_token"])
+    keyring.set_password(
+        DISPENSER_KEYRING_NAMESPACE, DISPENSER_KEYRING_REFRESH_TOKEN_KEY, token_data.get("refresh_token", "")
+    )
+    keyring.set_password(DISPENSER_KEYRING_NAMESPACE, DISPENSER_KEYRING_USER_ID_KEY, decoded_id_token.get("sub"))
+
+
+def clear_dispenser_credentials() -> None:
+    """
+    Clear the keyring passwords.
+    """
+
+    keyring.delete_password(DISPENSER_KEYRING_NAMESPACE, DISPENSER_KEYRING_ID_TOKEN_KEY)
+    keyring.delete_password(DISPENSER_KEYRING_NAMESPACE, DISPENSER_KEYRING_ACCESS_TOKEN_KEY)
+    keyring.delete_password(DISPENSER_KEYRING_NAMESPACE, DISPENSER_KEYRING_REFRESH_TOKEN_KEY)
+    keyring.delete_password(DISPENSER_KEYRING_NAMESPACE, DISPENSER_KEYRING_USER_ID_KEY)
+
+
+def is_authenticated() -> bool:
+    """
+    Check if the user is authenticated by checking if the token is still valid.
+    If the token is expired, attempt to refresh it.
+    """
+
+    try:
+        access_token = _get_auth_token()
+        rsa_pub_key = _get_access_token_rsa_pub_key(access_token)
+
+        jwt.decode(
+            access_token,
+            rsa_pub_key,
+            options={"verify_signature": True},
+            algorithms=ALGORITHMS,
+            audience=[
+                AUTH_CONFIG.audiences[DispenserApiAudiences.USER],
+                AUTH_CONFIG.audiences[DispenserApiAudiences.CI],
+            ],
+        )
+
+        return True
+    except jwt.ExpiredSignatureError:
+        logger.debug("Access token is expired. Attempting to refresh the token...")
+
+        try:
+            _refresh_user_access_token()
+            return True
+        except Exception:
+            logger.warning("Failed to refresh the access token. Retrying login...", exc_info=True)
```

Comment:
Does this actually retry the login? Or should the message be like below:
```suggestion
            logger.warning("Failed to refresh the access token. To login again, run `algokit dispenser login`", exc_info=True)
```

Reply:
I've set a slightly different message cause this actually is retrying login but only when invoked from within login command flow for any other case it will halt, so i just set a generic instruction in this text